prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,574 @@
+			entries[item].pid, name);
+		return -FI_EADDRINUSE;
+	}
+
+	/* fine, we could not find the entry, so now look for an empty slot */
+	for (item = 0; item < SM2_MAX_UNIVERSE_SIZE; item++) {
+		peer_pid = entries[item].pid;
+		if (peer_pid == 0)
+			goto found;
+		if (peer_pid < 0)
+			/* A third peer might have entered this address into
+			 * their AV, and there is no current way to check
+			 * this... need to keep this entry in the file until we
+			 * clean up */
+			continue;
+
+		if (!pid_lives(peer_pid)) {
+			sm2_file_extend_for_region(map, item);
+			entries = sm2_mmap_entries(map);
+			struct sm2_region *peer_region =
+				sm2_mmap_ep_region(map, item);
+
+			if (entries[item].startup_ready &&
+			    smr_freestack_isfull(sm2_free_stack(peer_region))) {
+				/* we found a slot with a dead PID and
+				 * the freestack is full */
+				entries[item].pid = 0;
+				goto found;
+			}
+		}
+	}
+
+	FI_WARN(&sm2_prov, FI_LOG_AV,
+		"No available entries were found in the coordination file, all "
+		"%d were used\n",
+		SM2_MAX_UNIVERSE_SIZE);
+	return -FI_EAVAIL;
+
+found:
+	if (self) {
+		entries[item].startup_ready = 0;
+		atomic_wmb();
+		entries[item].pid = pid;
+	}
+
+	if (!self && entries[item].pid == 0) {
+		entries[item].startup_ready = 0;
+		atomic_wmb();
+		entries[item].pid = -pid;
+	}
+
+	FI_WARN(&sm2_prov, FI_LOG_AV,
+		"Using sm2 region at allocation entry[%d] for %s\n", item,
+		name);
+
+	strncpy(entries[item].ep_name, name, SM2_NAME_MAX - 1);
+	entries[item].ep_name[SM2_NAME_MAX - 1] = '\0';
+
+	*gid = item;
+
+	sm2_file_extend_for_region(map, item);
+
+	return 0;
+}
+
+int sm2_entry_lookup(const char *name, struct sm2_mmap *map)
+{
+	struct sm2_ep_allocation_entry *entries;
+	int item;
+
+	entries = sm2_mmap_entries(map);
+	/* TODO Optimize this lookup*/
+	for (item = 0; item < SM2_MAX_UNIVERSE_SIZE; item++) {
+		if (0 == strncmp(name, entries[item].ep_name, SM2_NAME_MAX)) {
+			FI_DBG(&sm2_prov, FI_LOG_AV,
+			       "Found existing %s in slot %d\n", name, item);
+			return item;
+		}
+		if (entries[item].ep_name[0] != '\0') {
+			FI_DBG(&sm2_prov, FI_LOG_AV,
+			       "Searching for %s. Not yet found in spot "
+			       "allocation[%d]=%s, PID=%d\n",
+			       name, item, entries[item].ep_name,
+			       entries[item].pid);
+		} else if (!strcmp(entries[item].ep_name,
+				   ZOMBIE_ALLOCATION_NAME)) {
+			FI_DBG(&sm2_prov, FI_LOG_AV,
+			       "Found zombie in spot allocation[%d], PID= %d\n",
+			       item, entries[item].pid);
+		}
+	}
+	return -1;
+}
+
+/*
+ * Clear the pid for this entry.  must already hold lock.
+ */
+ssize_t sm2_entry_free(struct sm2_mmap *map, sm2_gid_t gid)
+{
+	struct sm2_ep_allocation_entry *entries;
+
+	entries = sm2_mmap_entries(map);
+	assert(entries[gid].pid == getpid());
+	entries[gid].pid = 0;
+	return 0;
+}
+
+ssize_t sm2_file_lock(struct sm2_mmap *map)
+{
+	struct sm2_coord_file_header *header = (void *) map->base;
+
+	pthread_mutex_lock(&header->write_lock);
+	return 0;
```

Comment:
There's no point in returning a value that's always 0.

Reply:
done